examples/src/ScalarDecoders.elm

```diff
@@ -0,0 +1,21 @@
+module ScalarDecoders exposing (Id, PosixTime, decoders)
+
+import Json.Decode as Decode exposing (Decoder)
+import Swapi.Scalar exposing (defaultDecoders)
+import Time
+
+
+type alias Id =
+    Swapi.Scalar.Id
+
+
+type alias PosixTime =
+    Time.Posix
+
+
+decoders : Swapi.Scalar.Decoders Id PosixTime
+decoders =
+    Swapi.Scalar.defineDecoders
+        { decoderId = defaultDecoders.decoderId
```

Comment:
this `defaultDecoders` business is tingling my spidey senses. To use a default, you have to know the default exists.

Would it be possible to require each scalar to have a separate module? Separating data structures is what modules do best! So then:

| Module | Has |
|-|-|
| `Scalar.Id` | `type alias Id`, `decode : Decoder Id` |
| `Scalar.PosixTime` | `type alias PosixTime`, `decode : Decoder PosixTime` |

Figuring out what default to set could be tricky, but maybe there could be a command-line flag for that? Or some GraphQL documentation annotation which generated the right thing?

Reply:
oh duh, easier way to do the right thing: write a default scalar value to the filesystem when it's missing and prompt the user to customize it. It's easy to assume they're all strings by default, and this way you can have some inline documentation on how to modify the code in place rather than having to copy or construct something from scratch.

Reply:
@BrianHicks thanks for the comments! I like the idea of having a file per scalar, that seems like an interesting path to explore.

Any ideas on how we could use a single-file-per-scalar approach and maintain these characteristics?

* The user should get a compiler error pointing to *their* custom scalar code (not the generated code) if they have a missing API.
* If a Scalar is removed or renamed, the user should get an error for referencing it (but if there's just a file that's not referenced anywhere they won't get an error)
* The code generator shouldn't have to touch or read files that the user created (this makes things way simpler to maintain and less bug-prone for the codegen, and is a nice sort of dependency inversion thing I think. I also think it leads to a simpler UX)

Reply:
> The user should get a compiler error pointing to their custom scalar code (not the generated code) if they have a missing API.

fill in a blank and tell them at generation time? That way if you delete a thing, it's kinda obvious what's going on. Not *as* good as the system you specified above but still maybe OK for other tradeoffy reasons like better organization.

> If a Scalar is removed or renamed, the user should get an error for referencing it (as opposed to just having a file that's not referenced anywhere and not getting an error)

Elm will generate an error for this, but not a good one for the use case, you're right.

> The code generator shouldn't have to touch or read files that the user created or changed (this makes things way simpler to maintain and less bug-prone for the codegen, and is a nice sort of dependency inversion thing I think. I also think it leads to a simpler UX)

What I'm proposing is a write-if-missing implementation, which means you'd reference stuff but not read/touch it from codegen otherwise.

---

Those feel like pretty weak answers to some good questions, which has revealed flaws! I'm still not 100% sure on lumping them into the same module but it seems *more* fine now. ;)